File: Lowest Common Ancestor of a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        def findNodePath(node, val, path):
            path.append(node)

            if node.val == val:
                return path

            if node.left is not None:
                leftPath = findNodePath(node.left, val, path)
                if leftPath is not None:
                    return leftPath

            if node.right is not None:
                rightPath = findNodePath(node.right, val, path)
                if rightPath is not None:
                    return rightPath

            path.pop()
            return None

        pPath = findNodePath(root, p.val, [])
        qPath = findNodePath(root, q.val, [])

        pSet = set(pPath)

        # print(pPath)
        # print(qPath)

        for i in range(len(qPath) - 1, -1, -1):
            if qPath[i] in pSet:
                return qPath[i]

File: Lowest Common Ancestor of a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    unordered_map<int, int> depth;
    unordered_map<int, TreeNode*> parents;

    int n;
    int max_depth = 0;
    int LOG;

    void dfs(TreeNode* parent, TreeNode* node, int d) {
        if (!node) return;

        parents[node->val] = parent;
        depth[node->val] = d;
        max_depth = max(max_depth, d);
        dfs(node, node->left, d+1);
        dfs(node, node->right, d+1);
    }

    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        dfs(root, root, 0);
        n = parents.size();

        int m2 = max_depth;
        LOG = 0;
        while (m2 != 0) {
            m2 >>= 1;
            LOG++;
        }

        // Preprocessing
        vector<unordered_map<int, TreeNode*>> up(LOG);
        for (auto [k, v]: parents) {
            up[0][k] = v;
        }

        for (int j = 1; j < LOG; j++) {
            for (auto [i, _]: parents) {
                up[j][i] = up[j-1][up[j-1][i]->val];
            }
        }

        if (depth[p->val] < depth[q->val]) {
            swap(p, q);
        }

        int k = depth[p->val] - depth[q->val];
        for (int j = 0; j < LOG; ++j) {
            if (k & (1 << j)) {
                p = up[j][p->val];
            }
        }

        if (p->val == q->val) return p;

        for (int j = LOG - 1; j >= 0; --j) {
            if (up[j][p->val] != up[j][q->val]) {
                p = up[j][p->val];
                q = up[j][q->val];
            }
        }
        return up[0][p->val];
    }
};
